src/nvim/api/extmark.c

```diff
@@ -431,6 +431,11 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id, Integer line, Integer
   }
 
   int line2 = -1;
+
+  if (HAS_KEY(opts->end_row) && !HAS_KEY(opts->end_line)) {
```

Comment:
I'd add an inner `if (HAS_KEY(opts->end_line)) { }` and return with an error if both keys are set.